### ISSUE TITLE
Porch: PackageVariant should set a readinessgate

### ISSUE THREAD

**Issue by liamfallon** [nephio-project/nephio#615]
Original issue URL: https://github.com/kptdev/kpt/issues/3979
Original issue user: https://github.com/johnbelamaric
Original issue created at: 2023-06-01T01:36:52Z
Original issue last updated at: 2023-07-06T21:56:34Z
Original issue body: ### Expected behavior
If PackageVariant fails to inject or otherwise properly mutate the Draft it creates, we need to prevent the package from being approved. This could be done by adding a False condition and readinessGate immediately after cloning (or better yet, implementing a porch server feature that allows that to be done atomically).

### Actual behavior
Because we cannot atomically "clone and apply PV mutations", the actual package revision process happens in two independent stages: clone, then apply mutations. If applying the mutations fails, the newly cloned Draft is still there, and unless you check the PV conditions, it looks fine and can be merged (in fact the auto-approval controller we built for Nephio goes ahead and approves it, since it only inspect the package revision, and doesn't know where it came from).

Original issue comments:
Comment user: https://github.com/johnbelamaric
Comment created at: 2023-07-06T21:56:34Z
Comment last updated at: 2023-07-06T21:56:34Z
Comment body: As a workaround for this, the approval controller now checks the PV readiness status, and so do some of our other controllers. But that wouldn't help a human user.

**Comment by liamfallon** [comment]
Triage Comments:
In the case of a PackageVariant, it will clone a package that has no readiness gate in it and it seems fine. Then as the package moves through the pipeline, there is no way to track whether of how much of the mutation in the pipeline has been executed. The PacakgeRevision creation is a 2 step process:
1. Clone the raw Package
2. Mutate the package with a pipeline or otherwise

The current implementation sets the package as ready after step 1.

To be considered in rearchitecture.

**Comment by JamesMcDermott** [comment]
## Some notes to be brought to Platform team

# Problem (this issue needs reframing):
  - a PackageRevision created by a PackageVariant should:
    - have a readiness gate and Condition (provisional type: PackagePipelinePassed) set False immediately after cloning
      - and the Condition updated to true once the package's pipeline completes successfully
    - so that while the pipeline is running, and if it fails, the PackageRevision is locked out from being proposed or approved
  - but there are other conditions when the pipeline might be run
    - for example, a human user using rpkg push to edit the PackageRevision's files
    - or a manual clone without a PackageVariant being involved at all
  - and there's not much apparent reason to limit the readiness-gate lockout to PackageVariant revisions
    - (or any easy way to automatically detect such a PackageRevision...)

# Therefore:
  - this PackagePipelinePassed readiness gate/Condition should NOT be set by the PackageVariant controller
  - instead, **PackageRevision conditions/readiness gates will be managed chiefly by the API server**
    - for this particular application, this means the server:
      - sets the PackagePipelinePassed readiness gate/condition upon clone/create - default False
      - catches the start of any pipeline runs to set the condition to False
      - catches a successful end of a pipeline run to set the condition to True
  - propose and approve operations will check all a PackageRevision's conditions in server during validation - logic:
        a.  get PackageRevision
        b.  read readiness gate list
        c.  read condition list, checking for conditions which match (a) readiness gate(s) (condition’s Type matches an entry in readiness gate list)
            i.	for each matching condition, check Status is True
            ii.	if False, break and fail
    - (validation in the porchctl propose/approve **already does this** – just need to migrate it down into the server to enforce it on API consumers as well)
  - **I propose this issue be renamed to "Porch: PackageRevision should set a readiness gate to control propose/approve"**

---

## Code impacts

### Already done:
- some work towards the original incorrect framing (managing Condition from PackageVariant controller)
  - see [PR #156](https://github.com/nephio-project/porch/pull/156)
  - notably:
    - discovered Porch stores readiness gates and Conditions in packages' Kptfiles
      - meaning any update to a Condition involves an additional resource update operation
      - which incurred extra work to make absolutely sure all code writes the files in the same way to avoid Git conflicts when applying the update
    - also adjusted update-package-revision logic to skip the pipeline if the update only sets readiness gates or conditions
      - allowing us to set readiness conditions based on pipeline success without triggering another pipeline run and getting into a loop

### To do:
- **rework previous work, move readiness gate and Condition management into API server**
  - somewhere in API server code - candidate locations:
    - generictaskhandler.go's applyResourceMutations() function
      - pro: common location where all mutations are run
      - con: it doesn't actually push the mutated changes, meaning it would need either a push from within applyResourceMutations() or separate pushes before/after each call of applyResourceMutations()
        - counterpoint: but it affects the PackageResources, which should theoretically allow us to pick out the Kptfile, check the renderStatus and update the PackagePipelinePassed condition right there, no need for a separate update push
          - not for setting the Condition to True on a successful pipeline, but still wouldn't handle a push beforehand to set the Condition to False
        - doable, yes... maintainable? Would need to try it and see
    - each of genericTaskHandler.ApplyTasks(), genericTaskHandler.DoPRMutations(), and genericTaskHandler.DoPRResourceMutations()
      - con: more places to handle the update
      - pro: but would probably have to handle them all anyway for the initial set-to-False push
      - pro: could wrap the invocations of applyResourceMutations() in a function - doWithReadinessCondition(func)?
      - as long as we're doing the update-to-True to the updated PackageResources that come out of a successful pipeline run, we should be good to push it all together to switch the Condition in the same operation
- **work out how to update a PackageRevision/PackageRevisionResources from within the server code to set (up) the readiness gate and Condition(s)**
  - route used in PackageVariant controller probably won't do since it was using the Kubernetes-shelled API operation
  - will require at least:
    - pick out the Kptfile from the package resources
    - parse to Porch Kptfile object if not already parsed
    - update gate/Condition as required
    - put it back in the package resources (writing to string if needed)
    - push to repo - necessary to make it available to external rpkg gets/describes!
- finish up migration of propose/approve validation into server
  - return informative error message if propose or approve fails - explain condition-based failure and which condition it was


## Open questions/pending decisions
- it's possible to slip in the crack between clone-package and update-to-set-readiness-gate and successfully propose/possibly approve it prematurely
  - as in, before the readiness gate/condition are set to False
  - however, I only got it using a while loop to run the porchctl propose constantly
  - this might widen a bit if there delays on the server - I tried a bit with long-running clone, but couldn't get one that ran long enough...
  - **question: would we be OK with that margin, or would it be better to update the create-package-revision flow to allow the readiness gates to be created out of the gate?**
- Greg raised one: **is Kubernetes condition-y logic and implementation usable in Porch as-is?**
  - just import/pull it in?
  - despite the different structure of the bespoke Porch Condition
    - extend the Kubernetes one?

### Test impacts
- e2e tests:
  - add capability to test PackageVariants
    - that was the original requirement of this issue
    - and it will still be useful for adding a long-running pipeline to a cloned downstream PackageRevision without having a long-running pipeline in the upstream that would slow down the clone operation
  - explore adding/editing a PackageRevision test to make sure of behaviour

### Remaining scenarios for manual testing
- test with delays in Packagerevision processing flow
  - between clone/update-readiness-gate
  - between pipeline end/update-readiness-gate-again
- Fiachra found an issue using the PackageVariant controller solution in the Free5GC test case
  - might be just sub-optimal solution - once reworked into server, need to ask him to try again
  - if issue appears again, need to reproduce Free5GC environment myself to debug - will likely need Fiachra's help to get past setup troubles

**Comment by JamesMcDermott** [comment]
Closing this for now - we'll revisit it once the PackageVariants are redesigned.
